### PR TITLE
Improve logic for showing dependency list: in progress, merged, deployed

### PR DESCRIPTION
**Pull request by marexandre**
screen shot 2016-02-10 at 5 40 54 pm

`Gray`: repo has commits
`Green`: ready to deploy, changes were merged into master but not deployed
`Strikethrough`: deployed

### CODE REVIEW

**Comment by MaximeVDG** [comment]
Visually LVGTM

File: static/js/pivotal.js [hunk 1]
```diff
@@ -143,21 +154,80 @@
...
       success: function(data) {
-        // Find all git commit messages
-        var list = data.filter(function(obj) {
-          return obj.commit_type;
-        });
-        // Find repo name from comment body text
-        list = list.map(function(item) {
-          return item.text.split('https://github.com/gengo/')[1].split('/commit/')[0];
-        });
+        var PULL_REQEST_REGEX = /Merge pull request/;
```
Comment (patatepartie) [inline static/js/pivotal.js]: `REQUEST` is missing `U`.

File: static/js/pivotal.js [hunk 2]
```diff
@@ -143,21 +154,80 @@
...
+        var COMMIT_REPO_REGEX = /https:\/\/github.com\/gengo\/(.+)\/commit\//;
+        var DEPLOY_REPO_REGEX = new RegExp('Deployed (.+) to '+ config.environment +': ');
+
+        var list = data.filter(function(activity) {
```
Comment (patatepartie) [inline static/js/pivotal.js]: `list` doesn't really tell us what this variable is about, except its type. Maybe `activities` would be a clearer name ?

Comment (marexandre) [inline static/js/pivotal.js]: :+1:

**Comment by marexandre** [comment]
@patatepartie Refactored the code a bit on your feedback, please take a look when you have time :santa:

File: static/js/pivotal.js [hunk 3]
```diff
@@ -143,21 +173,78 @@
...
-          return item.text.split('https://github.com/gengo/')[1].split('/commit/')[0];
+        var PULL_REQUEST_REGEX = /Merge pull request/;
+        var COMMIT_REPO_REGEX = /https:\/\/github.com\/gengo\/(.+)\/commit\//;
+        var DEPLOY_REPO_REGEX = new RegExp('Deployed (.+) to '+ config.environment +': ');
+
+        var activities = data.filter(function(activity) {
+          return activity.commit_type === 'github' || DEPLOY_REPO_REGEX.test(activity.text);
+        }).reverse();
+
+        var activitiesGrouped = groupBy(activities, function(activity) {
```
Comment (patatepartie) [inline static/js/pivotal.js]: I think it's valuable to add by what it's grouped so something like `activitiesGroupedByRepo`, or shorter: `activitiesByRepo`. What do you think ?

File: static/js/pivotal.js [hunk 4]
```diff
@@ -143,21 +173,78 @@
...
+          if (DEPLOY_REPO_REGEX.test(activity.text)) {
+            return activity.text.match(DEPLOY_REPO_REGEX)[1];
+          }
         });
 
-        callback(removeDupesFromArray(list));
+        var inProgress = [];
+        var deployed = [];
+        var notDeployed = [];
+        for (key in activitiesGrouped) {
```
Comment (patatepartie) [inline static/js/pivotal.js]: `Key` is actually the `repo`.

File: static/js/pivotal.js [hunk 5]
```diff
@@ -143,21 +173,78 @@
...
+          // Deployed message
+          if (DEPLOY_REPO_REGEX.test(activity.text)) {
+            return activity.text.match(DEPLOY_REPO_REGEX)[1];
+          }
         });
 
-        callback(removeDupesFromArray(list));
+        var inProgress = [];
+        var deployed = [];
+        var notDeployed = [];
```
Comment (patatepartie) [inline static/js/pivotal.js]: Maybe not so important, but I'm always wary of negatives. The brain plays tricks with them (like skipping the negation or struggling with double negatives).
What about `awaitingDeployment` or `readyToDeploy` or something ?

File: static/js/pivotal.js [hunk 6]
```diff
@@ -143,21 +173,78 @@
...
     });
   }
 
   /**
+   * getPopoverHTML return a HTML string
+   * @param  {Array} list Array of strings
+   * @return {String}     HTML text string
+   */
+  function getPopoverHTML(dependencies) {
+    // In progress
```
Comment (patatepartie) [inline static/js/pivotal.js]: If you name those variables with longer names, you won't need those comments.
Comments can easily become stale and then they can get confusing, so whenever I can I give priority to better names over more comments.

File: static/js/pivotal.js (hunk 5 above)
Comment (marexandre) [inline static/js/pivotal.js]: Will use `readyToDeploy`, thanks! :smile:

File: static/js/pivotal.js (hunk 4 above)
Comment (marexandre) [inline static/js/pivotal.js]: Thanks! Missed that :sleepy:

File: static/js/pivotal.js (hunk 3 above)
Comment (marexandre) [inline static/js/pivotal.js]: Like `activitiesByRepo`, will update the code! Thanks :beers:

File: static/js/pivotal.js (hunk 6 above)
Comment (marexandre) [inline static/js/pivotal.js]: True, some how was lazy to use longer names for variables... will update the code with proper variables names :)

**Comment by patatepartie** [comment]
Thanks for all those changes @marexandre : LGTM !

**Comment by marexandre** [comment]
@patatepartie Thanks for all the feedback! Had fun refactoring the code :monkey:

**Pull request merged**